### PR TITLE
Adding the non adversarial simulator

### PR DESCRIPTION
**Pull request by nagkumar91**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/evaluation/azure-ai-evaluation/README.md [hunk 1]
```diff
@@ -92,6 +92,125 @@ if __name__ == "__main__":
...
+    context: Optional[Dict[str, Any]] = None,
+) -> dict:
+    messages_list = messages["messages"]
+    # get last message
+    latest_message = messages_list[-1]
+    query = latest_message["content"]
+    context = None
+    # call your endpoint or ai application here
+    current_dir = os.path.dirname(__file__)
+    prompty_path = os.path.join(current_dir, "application.prompty")
```
Comment (w-javed) [inline sdk/evaluation/azure-ai-evaluation/README.md:167]: We should wrap around the prompty part in the separate function to flow the pattern given in [here](https://learn.microsoft.com/en-us/azure/ai-studio/how-to/develop/simulator-interaction-data#specify-target-callback-to-simulate-against)

Comment (nagkumar91) [reply sdk/evaluation/azure-ai-evaluation/README.md]: done!

**Review by changliu2 — COMMENTED** [review]
first pass on task_simulate

File: sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/synthetic/simulator.py [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+    ) -> List[JsonLineChatProtocol]:
+        """
+        Generates synthetic conversations based on provided parameters.
+
+        :keyword target: The target function to call during the simulation.
+        :paramtype target: callable
+        :keyword max_conversation_turns: Maximum number of conversation turns for the simulation.
+                                        Each turn consists of a user and an assistant message.
+        :paramtype max_conversation_turns: int
+        :keyword tasks: A list of user tasks, each represented as a list of strings.
```
[the remark this answers is not in the document]
Comment (nagkumar91) [reply sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/synthetic/simulator.py]: Feel free to add more documentation explaining `how many turns will be generated` The concept of `turn` has been pretty vague with some expecting a query, response pair to be a turn vs others saying a query and a response is 2 turns.

If len(tasks) > num_queries, the first num_queries of tasks will be executed. I can add a message to the user warning about this. Happy to get any suggestions on how you would want to phrase it.

`num_queries` is used only for the first pass of the simulator where query and expected responses are generated. If `conversation_turns` is passed, the `num_queries` is not needed as its not used. 
`tasks` are passed on as arguments to the prompty that simulates the user.
for ex. 
```
max_conversation_turns = 3
task = 'I need to finish my homework'
# computed query response 
{
    "query": "what is the capital of france?",
    "response": "france"
}
```
In this situation, the conversation history would look like:
```
UserMessage("what is the capital of france?")
AssistantMessage(<whatever the callback responds with>)
UserMessage(<follow up generated by the prompty which is trying to accomplish task>)
AssistantMessage(<whatever the callback responds with>)
UserMessage(<follow up generated by the prompty which is trying to accomplish task>)
AssistantMessage(<whatever the callback responds with>)
```

File: sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/synthetic/simulator.py [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+        :paramtype user_simulator_prompty: Optional[str]
+        :keyword api_call_delay_sec: Delay in seconds between API calls.
+        :paramtype api_call_delay_sec: float
+        :keyword query_response_generating_prompty_kwargs: Additional keyword arguments for the query response generating prompty.
+        :paramtype query_response_generating_prompty_kwargs: Dict[str, Any]
+        :keyword user_simulator_prompty_kwargs: Additional keyword arguments for the user simulator prompty.
+        :paramtype user_simulator_prompty_kwargs: Dict[str, Any]
+        :keyword conversation_turns: Predefined conversation turns to simulate.
+        :paramtype conversation_turns: List[List[str]]
+        :return: A list of simulated conversations represented as JsonLineChatProtocol objects.
```
[the remark this answers is not in the document]
Comment (nagkumar91) [reply sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/synthetic/simulator.py]: `context` is something that is present in the message protocol. Adding a new key would make it harder for the evaluation step as it is not expecting those new keys. Doing so might add a translation step which combines them back into context to help evaluation.

File: sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/synthetic/simulator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (changliu2) [reply sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/synthetic/simulator.py]: Suggested documentation:

**docstring for simulator**: this function supports two modes: 
1) task-free mode: when only 'num_queries' is specified and 'tasks' is not, return a jsonl output where each line is a turn of conversation grounded in the context of the 'text' and there will be in total 'num_queries' x 'max_conversation_turns' lines of simulated data.
2) task-specific mode: when both 'num_queries' and 'tasks' are specified, then only max('num_queries', len('tasks')) x 'max_conversation_turns' lines of simulated data grounded in the end-user scenario specified by each item in 'tasks'. In particular, when 'num_queries' > len('tasks'), all tasks will be used for generation and the remaining 'num_queries' - len('tasks') lines will be simulated in task-free mode. When 'num_queries' < len('tasks'), only the first 'num_queries' lines of the specified tasks will be simulated.

**warning message**: When 'num_queries' > len('tasks'), output warning("You have specified 'num_queries' > len('tasks'). All tasks will be used for generation and the remaining 'num_queries' - len('tasks') lines will be simulated in task-free mode"). When 'num_queries' < len('tasks'), output warning("You have specified 'num_queries' < len('tasks'). Only the first 'num_queries' lines of the specified tasks will be simulated.

btw, does max_conversation_turns suggest early stop or is the results guaranteed to have num_queries x max_conversation_turns lines? @nagkumar91

**Review by changliu2 — APPROVED** [review]
resolved issues after improving docstring and cut down output paths to simplify.

**Review by changliu2 — APPROVED** [review]
resolved issues by improving docstring/warning messages and cutting down possible output paths to simplify.

**Review by luigiw — APPROVED** [review]

**Pull request merged**